Load form submissions from loadForm page.

Stni18 opened the pull request:
I am open to suggestions on how I can better format the content that is being loaded from loadFormData.js and other changes I should make. [loadPage.html] works as intended; however, they show case "," in the text which can be bit distracting.

logannc commented on portfolio/src/main/java/com/google/sps/servlets/LoadFormServlet.java:
```diff
@@ -0,0 +1,76 @@
+      String subject = entity.getString("subject");
+      String comment = entity.getString("text-input");
+
+      String[] content = {name, " " + subject, " " + comment, "\n\n" };
```
What is the `" " + var` for?

logannc replied:
Do you want this to just be a list of lists or did you want it to be structured?

Stni18 replied:
I was thinking of including a list that would display the specified entities, but I was curious to see if I could structure them in the following way: 

loadForm.html : (What I was trying to do)

Name: (name entity)     Subject: (subject entity)    Comment: (Comment entity), but when it prints out the values (or rather, displays the values in the html page)  they are separated by "," which at times can make it a little hard to make out which entity is which, this is  why I added the "/" in the commented code and at the paragraph to match it.  Hope this gave a bit more context; nonetheless, I will be making the recommended changes.

Stni18 replied:
Also, the (" ") before the variable was to leave a bit of space between the commas when the entities would be loaded.

logannc replied:
That shouldn't matter if you actually use the json instead of just serializing it

logannc commented on portfolio/src/main/java/com/google/sps/servlets/LoadFormServlet.java:
```diff
@@ -0,0 +1,76 @@
+    //   forms.add(" " + "/"+ timeStamp);
+    //   //forms.add(" " + email);
+    //   forms.add(" "+ "/"+ subject);
+    //   forms.add(" " + "/"+ comment + "\n\n");
```
remove commented out code

logannc commented on portfolio/src/main/webapp/formEmailNotice.js:
```diff
@@ -0,0 +1,17 @@
+    }
+});
+
+
```
format doc/remove empty space

logannc commented on portfolio/src/main/webapp/loadForm.html:
```diff
@@ -0,0 +1,52 @@
+
+        <form action="/load-form" method="GET" >
+
+            <p><span style = "font-weight:bold"> [/Name, /Subject, /Comments] </span></p>
```
whats with the slashes?

logannc commented on portfolio/src/main/webapp/loadForm.html:
```diff
@@ -0,0 +1,52 @@
+
+            <p><span style = "font-weight:bold"> [/Name, /Subject, /Comments] </span></p>
+            <p><span style="color:red; font-weight:bold"> *Emails are not displayed for the security of the clients*
+            </span></p>
```
format doc

logannc commented on portfolio/src/main/webapp/loadForm.html:
```diff
@@ -0,0 +1,52 @@
+
+        <h1 class="title"> Submissions: </h1>
+
+        <form action="/load-form" method="GET" >
```
why do you have a form?

logannc commented on portfolio/src/main/webapp/style.css:
```diff
@@ -93,11 +93,10 @@ iframe {
+	width: 100%;
 	margin-right: 40px;
-    margin-left: 20px;
+    margin-left: 40px;
```
format doc

logannc commented on portfolio/src/main/webapp/loadFormData.js:
```diff
@@ -0,0 +1,28 @@
+
+//Fetches data from form and parses it to HTML as a list that loads all of the messages
+
+async function loadForm() {
```
this seems to not be called anywhere?

Stni18 replied:
I called it in the body of the loadForm.html page, using on onload = "loadForm()"

logannc replied:
You should, generally, not use javascript directly in HTML.

There are a couple options here.

1. You can just call the function in your javascript file. This will make the function run but it has drawbacks if the page isn't fully loaded.
2. https://developer.mozilla.org/en-US/docs/Web/API/Document/DOMContentLoaded_event#basic_usage
3. https://developer.mozilla.org/en-US/docs/Web/API/Document/readyState#readystatechange_as_event_listener_to_insert_or_modify_the_dom_before_domcontentloaded

For 2/3, you would add the event listender to check when the page is sufficiently loaded, then you would call your function to kick off the request/DOM manipulation.

Stni18 replied:
So I should use document.body.onload = loadForm? I tried that once and it didn't work

logannc replied:
I mean that should work, I guess, but that wasn't one of the three options above

logannc commented on portfolio/src/main/webapp/loadFormData.js:
```diff
@@ -0,0 +1,28 @@
+    // reference to create HTML.
+    const myForm = await responseFromServer.json();
+
+    formContainer.innerText = myForm;
```
I think the intent of this step in the project is to do more than just `container.innerText = blah`

process the json input and add elements e.g. https://developer.mozilla.org/en-US/docs/Web/API/Document/createElement

slightly more applicable example (that I have just written on the fly, without testing so it may not work or match your exact structure):

```javascript
let docContainer = document.getElementById('form-container');
let listContainer = document.createElement('ul');
for (var i = 0; i < list.length; i++) {
    let submission = list[i];
    let elem = document.createElement('li');
    elem.appendChild(document.createTextNode(`${submission.name} (${submission.subject}): ${submission.comment}`));
    listContainer.appendChild(elem);
}
docContainer.appendChild(listContainer);
```

Stni18 replied:
Thanks, I'll read the documentation and test out the example. Should I have any doubts I'll check in with the Technical Project Chat.

logannc commented on portfolio/src/main/java/com/google/sps/servlets/LoadFormServlet.java:
```diff
@@ -0,0 +1,54 @@
+            String subject = entity.getString("subject");
+            String comment = entity.getString("text-input");
+
+            String[] content = { "Name: " + name, " Subject: " + subject, " Comments: " + comment };
```
Do the formatting client side.

Stni18 replied:
I'm not sure what you mean by formatting it in the client side, Do you mean I should add the "Name: " , "Subject: " and "Comment: " manually  to my loadForm.html page?  I have tried that, but all of the content is displayed by using a <p> </p> that contains the id of the loadForm() function in my javaScript.  I also tried different ways to format it from my javaScript, but was not able to.

logannc replied:
Consider this: if I asked you to put each item in their own `span` (e.g., `<span><b>Name:</b>Some Example Name</span>`), how would you do that?

Your goal here is to, using only *data* from the server, generate markup on the client.

logannc commented on portfolio/src/main/webapp/loadFormData.js:
```diff
@@ -0,0 +1,36 @@
+    console.log("Submissions:", myForm);   //This works
+
+
+    const list = myForm; //const variable that stores the submissions as list
```
why do you reassign?

Stni18 replied:
If I didn't reassign it would give a error saying that the list[i] was not found inside the function.

Stni18 replied:
const list = myForm; //const variable that stores the submissions as list

    //Displays list of form submissions in loadForm.html
    let docContainer = document.getElementById('form-container');
    let listContainer = document.createElement('ul');
    for (var i = 0; i < list.length; i++) {
        let submission = list[i];
        let elem = document.createElement('li');
        elem.appendChild(document.createTextNode(`${submission}`));
        listContainer.appendChild(elem);
    }
    docContainer.appendChild(listContainer);               This should give a better context of what I mean

logannc replied:
I'm aware of the literal reason, but you should probably just change the loop to reference the variables original name or something instead of just giving the same value different names.

Stni18 replied:
Ok, so if I just rename the object to "list" or rename the interior to "myForm" I'm done with this portion? Just want to make sure, before I modify and commit again.

logannc commented on portfolio/src/main/webapp/loadForm.html:
```diff
@@ -30,12 +30,18 @@
-        <p id="form-container"></p>
+        <table>
+            <tr>
+				<td><h3>&emsp; Name &emsp; &nbsp;</h3></td>
```
Don't use spaces like this. If you're trying to center the text, use CSS.

logannc commented on portfolio/src/main/webapp/loadFormData.js:
```diff
@@ -25,12 +25,43 @@ async function loadForm() {
+    //------------------------------------
+
+    // creates a <table> element and a <tbody> element
+    var tbl = document.createElement("table");
```
instead of creating a new table, you could get the table that already exists?

Stni18 commented:
Regarding the latest commit I forgot to mention that I only did the pie chart portion until the hard coded portion (step 2/7, the walkthrough asked me to commit at this portion anyways).

logannc commented on portfolio/src/main/java/com/google/sps/servlets/LoadFormServlet.java:
```diff
@@ -0,0 +1,54 @@
+                .build();
+        QueryResults<Entity> results = datastore.run(query);
+
+        ArrayList<Object> forms = new ArrayList<>();
```
Just noticed this. Seeing `Object` is generally a bad sign. If you make `content` an `ArrayList<String>`, then this can be

```suggestion
        ArrayList<ArrayList<String>> forms = new ArrayList<>();
```

logannc commented on portfolio/src/main/webapp/loadFormData.js:
```diff
@@ -0,0 +1,62 @@
+
+    const responseFromServer = await fetch('/load-form');
+
+    const formContainer = document.getElementById('form-container');
```
You seem to get this element twice here and on line 27

logannc commented on portfolio/src/main/webapp/loadFormData.js:
```diff
@@ -0,0 +1,62 @@
+    //------------------------------------
+
+    // creates a <table> element and a <tbody> element
+    var tbl = document.getElementById("table");
```
I don't see an element with id `table`.

Stni18 commented:
My bad,  I accidently hit crtl + z before committing the document to GitHub, the id should now be in the loadForm.html document.

logannc commented:
No worries. Deal with these last few concerns and format all of the files in this PR and you'll be good to go!

Stni18 commented:
I believe to have now addressed all of the issues, should there be anything else I should add please let me know. I addition, I am curious as to why **Object** was a bad sign in the ArrayList<>. Why is that so?

logannc commented:
Because you're throwing away type information. Why call it an Object when we know its an array? When you cast to Object, we can no longer treat it as the type that it is. We'd have to manually cast it back to its proper type later.

It isn't a big deal here because of the small scope, but as a general rule don't use Object unless there are not other alternatives.
